### PR TITLE
48 add-translate-client-cli-tool

### PR DESCRIPTION
**Pull request by klavinsjanis**

### CODE REVIEW

File: cli/translate/translate
Comment (janishorsts) [inline cli/translate/translate:1]: compiled file in repo? :)

File: cli/translate/cmd/root.go
Comment (janishorsts) [inline cli/translate/cmd/root.go:1]: We keep all CLI commands under `cmd`.

File: cli/translate/translate
Comment (klavinsjanis) [reply cli/translate/translate:1]: ready to use cli tool :)

File: cli/translate/cmd/service/ls.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+			grpc.WithBlock(),
+		}
+
+		// Wait for the server to start and establish a connection.
+		conn, err := grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			log.Panicf("list services: connect to service: %v", err)
+		}
+
+		client := translatev1.NewTranslateServiceClient(conn)
```
[the remark this answers is not in the document]
Comment (klavinsjanis) [reply cli/translate/cmd/service/ls.go]: I believe I could establish a default client, but if someone will pass in new address, I will have to rebuild the client for the command in question.

File: cli/translate/cmd/root.go
Comment (klavinsjanis) [reply cli/translate/cmd/root.go:1]: yeah, kinda figured that was the best way and refactored.

File: cli/translate/cmd/service/ls.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		t.AppendFooter(table.Row{"", "Total", len(resp.Services)})
+		t.SetStyle(table.StyleLight)
+		t.Render()
+	},
+}
+
+func init() {
+	flags := lsCmd.Flags()
+
+	flags.StringP("service.address", "a", "localhost:8080",
```
[the remark this answers is not in the document]
Comment (klavinsjanis) [reply cli/translate/cmd/service/ls.go]: changed it to address, just copied it from the cmd/translate

File: cmd/client/cmd/service.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	ctx, cancelFunc := context.WithTimeout(ctx, time.Second*10) //nolint:gomnd
+	defer cancelFunc()
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+	}
+
+	// Wait for the server to start and establish a connection.
+	return grpc.DialContext(ctx, address, opts...) //nolint:wrapcheck
```
Comment (VladislavsPerkanuks) [inline cmd/client/cmd/service.go:92]: Do we use linters to ignore them? :D
Why not wrap 🤔

File: cmd/client/main.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"os"
+
+	"go.expect.digital/translate/cmd/client/cmd"
+)
+
+func main() {
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
```
Comment (janishorsts) [inline cmd/client/main.go:11]:
```suggestion
		log.Panic(err)
```

File: cmd/client/cmd/service.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	translatev1 "go.expect.digital/translate/pkg/pb/translate/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func init() {
+	flags := lsCmd.Flags()
+
+	flags.StringP("address", "a", "localhost:8080",
+		"address for the translate agent GRPC client")
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:21]:
```suggestion
		`"translate" service address as "host:port"`)
```

File: cli/translate/main.go
Comment (VladislavsPerkanuks) [inline cli/translate/main.go:1]: Forgot to delete ?  cli/translate is the same as cmd/client

File: cmd/client/cmd/service.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		ctx := context.Background()
+
+		client, err := newClientConn(ctx, addr)
+		if err != nil {
+			log.Panicf("list services: new GRPC client connection: %v", err)
+		}
+
+		resp, err := translatev1.NewTranslateServiceClient(client).ListServices(ctx, &translatev1.ListServicesRequest{})
+		if err != nil {
+			log.Panicf("list services: make GRPC request: %v", err)
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:60]:
```suggestion
			log.Panicf("list services: send GRPC request: %v", err)
```

File: cli/translate/main.go
Comment (klavinsjanis) [reply cli/translate/main.go:1]: broken git :D

File: cli/translate/cmd/service.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+		t.AppendFooter(table.Row{"", "Total", len(resp.Services)})
+		t.SetStyle(table.StyleLight)
+		t.Render()
+	},
+}
+
+// helpers
+
+func newClientConn(ctx context.Context, address string) (*grpc.ClientConn, error) {
+	ctx, cancelFunc := context.WithTimeout(ctx, time.Second*10) //nolint:gomnd
```
Comment (janishorsts) [inline cli/translate/cmd/service.go]: Use  the viper `timeout` argument

File: cli/translate/cmd/service.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	defer cancelFunc()
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
```
Comment (janishorsts) [inline cli/translate/cmd/service.go]: use the viper `insecure` (bool) argument to decide if to use insecure or secure connection (default).

File: cmd/client/cmd/service.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		log.Panicf("bind timeout flag: %v", err)
+	}
+
+	serviceCmd.AddCommand(lsCmd)
+}
+
+// serviceCmd represents the service command.
+var serviceCmd = &cobra.Command{
+	Use:   "service",
+	Short: "Service holds list of commands for interacting with services.",
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:44]:
```suggestion
	Short: "Manage services",
```

File: cmd/client/cmd/service.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+		if err := cmd.Help(); err != nil {
+			log.Panicf("display help: %v", err)
+		}
+	},
+}
+
+// lsCmd represents the lsCmd command.
+var lsCmd = &cobra.Command{
+	Use:   "ls",
+	Short: "Lists services.",
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:55]: Let's avoid `.` in the short descriptions.

```suggestion
	Short: "Lists services",
```

File: cmd/client/cmd/service.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+const establishConnTimeout = 5 * time.Second
+
+func init() {
+	flags := lsCmd.Flags()
+
+	flags.StringP("address", "a", "localhost:8080", `"translate" service address as "host:port"`)
+	flags.BoolP("insecure", "i", false, `disable transport security (default false)`)
+	flags.DurationP("timeout", "t", establishConnTimeout, `timeout for establishing connection with "translate" service`)
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:24]: `timeout` limits the time CLI execution time. Please take it into account.

File: cmd/client/cmd/service.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+	if err := viper.BindPFlag("address", flags.Lookup("address")); err != nil {
```
Comment (janishorsts) [inline cmd/client/cmd/service.go:26]: Can we use viper unmarshal to reduce the code boilerplate?

File: cmd/client/main.go
Comment (janishorsts) [inline cmd/client/main.go:1]: We require tests

Comment (klavinsjanis) [reply cmd/client/main.go:1]: ok

**Pull request closed without merging**